File: whatmakesaman/3주차/PG_67256_키패드_누르기.py
position={
    '1': (0,0), '2': (1,0), '3': (2,0),
    '4': (0,1), '5': (1,1), '6': (2,1),
    '7': (0,2), '8': (1,2), '9': (2,2),
    '*': (0,3), '0': (1,3), '#': (2,3),
}

def distance(a,b):
    x1, y1=position[a]
    x2,y2=position[b]
    
    return abs(x1-x2)+abs(y1-y2)

def solution(numbers, hand):
    answer = ''
    left, right='*','#'
        
    for n in numbers:
        s=str(n)      
        
        if s in ('1','4','7'):
            answer+='L'
            left=s
            
        elif s in ('3','6','9'):
            answer+='R'
            right=s
            
        else:
            dl=distance(s,left)
            dr=distance(s, right)
            
            if dl < dr:
                answer+='L'
                left=s
            elif dl > dr:
                answer+='R'
                right=s
            else:
                if hand=='left':
                    answer+='L'
                    left=s
                else:
                    answer+='R'
                    right=s
    return answer

File: kitttae/3주차/PG_67256_키패드_누르기.java
class Solution {
    
    int[][] numPhone = {
        {3,1},
        {0,0},
        {0,1},
        {0,2},
        {1,0},
        {1,1},
        {1,2},
        {2,0},
        {2,1},
        {2,2},
        {3,0},
        {3,2}
    };
    
    public String solution(int[] numbers, String hand) {
        StringBuilder sb = new StringBuilder();
        int left = 10;
        int right = 11;
        
        for (int num : numbers) {
            if (num == 1 || num == 4 || num == 7) {
                sb.append("L");
                left = num;
            }
            else if (num == 3 || num == 6 || num == 9) {
                     sb.append("R");
                     right = num;
            }
            else {
                int leftL = getDistance(left, num);
                int rightR = getDistance(right, num);
                
                if (leftL < rightR) {
                    sb.append("L");
                    left = num;
                }
                else if (rightR < leftL) {
                    sb.append("R");
                    right = num;
                }
                else {
                    if (hand.equals("left")) {
                        sb.append("L");
                        left = num;
                    }
                    else {
                        sb.append("R");
                        right = num;
                    }
                }
            }
        }
        
        return sb.toString();
    }
    
    private int getDistance(int handnum, int n) {
        int r1 = numPhone[handnum][0];
        int c1 = numPhone[handnum][1];
        int r2 = numPhone[n][0];
        int c2 = numPhone[n][1];
        
        return Math.abs(r1 - r2) + Math.abs(c1 - c2);
    }
    
    
}
